Raise alignment limit from 2^15 to 2^31 - 1

PlasmaPower opened the pull request:
Fixes #42960

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @arielb1 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

PlasmaPower commented:
@eddyb - I think you wrote the initial Align struct in #32939, does this PR look good?

I don't think I can change the reviewer after creating the PR.

eddyb commented:
@PlasmaPower Did you mean `2^64 - 1`? Or are you bounding it somewhere?

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -312,39 +312,36 @@ impl Align {
         Ok(Align {
-            raw: pack(abi)? | (pack(pref)? << 4)
+            abi: pack(abi)?,
+            pref: pack(pref)?,
```
I'd rename this to `log2`.

eddyb commented on src/libsyntax/attr.rs:
```diff
@@ -974,11 +974,11 @@ pub fn find_repr_attrs(diagnostic: &Handler, attr: &Attribute) -> Vec<ReprAttr>
-                                // rustc::ty::layout::Align restricts align to <= 32768
-                                if align <= 32768 {
-                                    acc.push(ReprAlign(align as u16));
+                                // rustc::ty::layout::Align restricts align to <= 2147483648
```
Ah but this is incorrect given that `Align::from_bytes` takes an `u64`.

PlasmaPower commented:
@eddyb you're right, I did the math on the upper bound wrong. Will fix.

PlasmaPower commented:
The resulting binary is segfaulting when I switch from `u32`s to `u64`s in some places, as necessary with a limit of `2^64 - 1`. I'll investigate further.

eddyb commented:
@PlasmaPower LLVM probably doesn't like alignments larger than `2^31 - 1`.

PlasmaPower commented:
@eddyb I'm assuming you mean larger than `2^32 - 1`? I'm not sure why that would matter, and in addition I'm not testing alignments larger than `2^32 - 1`. There might be another factor here. I might just stick with `u32`s though, as we shouldn't need anything larger than that.

eddyb commented:
@PlasmaPower LLVM uses signed 32-bit integers for alignments.

PlasmaPower commented:
I rebased on master and added the limit changes, but resulting binaries with big aligns still seem to be segfaulting. They were working before, but when I check out the previous commit it still seg faults. I might try removing my build directory and building the compiler fresh. Any ideas @eddyb ? When I run it with GDB, it says "no stack" after segfaulting.

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -312,39 +314,36 @@ impl Align {
             if bytes != 1 {
                 Err(format!("`{}` is not a power of 2", align))
-            } else if pow > 0x0f {
-                Err(format!("`{}` is too large", align))
```
You should keep the check here though.

PlasmaPower replied:
Should this struct care about the 31 bit limit? I've done the checking for that closer to the parsing. The previous limit here was due to the limitations of using just 1 nibble, but now it has no such limit.

eddyb replied:
The parsing is for the user-facing value. Here we must protect any internal guarantees.

PlasmaPower replied:
Fair enough. Will fix.

eddyb commented on src/test/run-pass/align-struct.rs:
```diff
@@ -193,4 +200,14 @@ pub fn main() {
+    arr[0] = 132;
+    let large = AlignLarge {
+        stuff: arr,
+    };
```
Try using `box` (not `Box::new`!) - maybe you're blowing the stack?

PlasmaPower replied:
Good point, I thought an upstream change might result in that but I didn't think to avoid it with `box`.

eddyb commented:
cc @alexcrichton The PR in its current state might be failing in strange ways due to the new probestack logic. For the record, [the updated test](https://github.com/PlasmaPower/rust/blob/a10213f297be8e231f55fd6db843674a8f512140/src/test/run-pass/align-struct.rs#L204-L212) contains:
```rust
// The align limit was originally smaller (2^15).
// Check that it works with big numbers.
#[repr(align(0x10000))]
struct AlignLarge {
    stuff: [u8; 0x10000],
}

// Check the actual address is aligned
fn is_aligned_to<T>(p: &T, align: usize) -> bool {
    let addr = p as *const T as usize;
    (addr & (align - 1)) == 0
}

fn main() {
    let mut arr = [0; 0x10000];
    arr[0] = 132;
    let large = AlignLarge {
        stuff: arr,
    };
    assert_eq!(large.stuff[0], 132);
    assert_eq!(mem::align_of::<AlignLarge>(), 0x10000);
    assert_eq!(mem::align_of_val(&large), 0x10000);
    assert!(is_aligned_to(&large, 0x10000));
}
```

PlasmaPower commented:
@eddyb That's exactly the change I thought of, but I don't know near enough about that low level stuff to know how to work around it or fix it.

PlasmaPower commented:
Box syntax fixed it, but it turns out it wasn't the large alignment but instead the large array. This used to work: `let x = [0; 0x10000]` but now it needs to be `let x = box [0; 0x10000]`.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit b4973e9 has been approved by `eddyb`

<!-- @bors r=eddyb b4973e961974ac9bdd9076acbe783b0e43242256 -->

bors commented:
:hourglass: Testing commit b4973e961974ac9bdd9076acbe783b0e43242256 with merge c531c46140a9d65da3e2d24396ebde3ee1c54335...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/251552722?utm_source=github_status&utm_medium=notification)

PlasmaPower commented:
Failure seems unrelated to this PR.

eddyb commented:
@bors retry

bors commented:
:hourglass: Testing commit b4973e961974ac9bdd9076acbe783b0e43242256 with merge 4b7f41a22509d87a26d429185aeeaba0a2d024c5...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3850), [status-travis](https://travis-ci.org/rust-lang/rust/builds/251581253?utm_source=github_status&utm_medium=notification)
Approved by: eddyb
Pushing 4b7f41a22509d87a26d429185aeeaba0a2d024c5 to master...
